### DISCUSSION TITLE
Possible to always assign a new (incremented) virtual ip from the pool for the same IKE ID?

### DISCUSSION THREAD

**Discussion by thomas-strangert-emblasoft** [strongswan/strongswan#2570] [Q&A] [answered]
Is there some setting that makes the virtual ip pool handler become "less clever" and instead always hand out a new / incremented ip address when the same IKE identity wants to initiate a new connection?
As it is now, a previously connected user will get the same vip from the pool.

The swanctl setting <conn>.unique=never/no/keep/replace is sort of in that area, but has afaik more to do with how to allow new/kick out previous SAs for the same user, not how to allocate a new/find a previosly used vip.

My issue is that if a user drops out and then reconnects/re-authenticates and gets the same client vip, then some other nodes on the "inside LAN" can get a bit confused and not challenge the session as "new".

**Comment by tobiasbrunner** [comment, marked as answer]
> Is there some setting that makes the virtual ip pool handler become "less clever" and instead always hand out a new / incremented ip address when the same IKE identity wants to initiate a new connection?

No, that's currently not possible (it could also be problematic if reauthentication is used).

> My issue is that if a user drops out and then reconnects/re-authenticates and gets the same client vip, then some other nodes on the "inside LAN" can get a bit confused and not challenge the session as "new".

Hm, not sure I get what you mean. What difference does it make if the IP is reused by the the same or a different user? Wouldn't the latter be even worse?

**Reply by thomas-strangert-emblasoft** [reply]
> > Is there some setting that makes the virtual ip pool handler become "less clever" and instead always hand out a new / incremented ip address when the same IKE identity wants to initiate a new connection?
> 
> No, that's currently not possible (it could also be problematic if reauthentication is used).
> 
Oki, "currently not possible". Too bad for me then. (I didn't think about possible reauthentication issues since my config doesn't use it, only rekeying.)

> > My issue is that if a user drops out and then reconnects/re-authenticates and gets the same client vip, then some other nodes on the "inside LAN" can get a bit confused and not challenge the session as "new".
> 
> Hm, not sure I get what you mean. What difference does it make if the IP is reused by the the same or a different user? Wouldn't the latter be even worse?
> 
A bit simplified, but to give you an idea: The protocol that is used over the ipsec include a user token so it is the token+ip combination that gives how the application will detect/challenge the connection as new or already existing. 
While this mechanism in no way is perfect, it would however work like 100% if a new ipsec session for a user would get a new ip rather than that the user reuses the ip from the previous ipsec session. (A new ipsec session with a new ip will force a challenge from the application.)
If a different user gets an "already used" ip, it will also be detected and challenged since that user would have a different user token, so the token+ip combination will be new.
It is the new ipsec session with same/reused ip + same user token that is a bit more tricky for the application to detect as "new", since it has the same token+ip as the previous ipsec session.

**Reply by tobiasbrunner** [reply]
Maybe you could use the [`dhcp` plugin](https://docs.strongswan.org/docs/latest/plugins/dhcp.html). If the `identity_lease` option is disabled, a random MAC address is sent to the DHCP server for every request, so that might allow you to achieve this if the DHCP server has an option to first create new leases instead of reusing old ones. RADIUS provides another external option to allocate virtual IP. And you could also write your own plugin I guess.



> It is the new ipsec session with same/reused ip + same user token that is a bit more tricky for the application to detect as "new", since it has the same token+ip as the previous ipsec session.

Why does it have to be detected as new? I mean it's the same token and IP.

**Reply by thomas-strangert-emblasoft** [reply]
> Maybe you could use the [`dhcp` plugin](https://docs.strongswan.org/docs/latest/plugins/dhcp.html). If the `identity_lease` option is disabled, a random MAC address is sent to the DHCP server for every request, so that might allow you to achieve this if the DHCP server has an option to first create new leases instead of reusing old ones. RADIUS provides another external option to allocate virtual IP. And you could also write your own plugin I guess.
> 
Yes, I have contemplated to use the dhcp plugin as an alternative if it turned out (which it has) that the vip pool handling couldn't be forced to always do ip increments.
I wonder though how the dhcp lease period is periodically renewed/maintained, given that my ipsec clients may have long session times and don't do reauthentication, only rekeying? I suppose that the dhcp plugin will only ask the dhcpd server for an initial ip address (with DHCPDISCOVER) when the client asks for an ip address, and that won't happen upon rekeying only. Or will it?
The dhcpd server I would use is linux/Ubuntu standard ICS DHCP 4.4 and it has certain methods (e.g. ping) when handing out ips in order to check if an ip in its pool range is in fact used by someone even though the ip isn't internally marked as being allocated. This might actually suffice to avoid ip allocation conflicts if the client/dhcp plugin doesn't renew its dhcp lease (with DHCPREQUEST).

> > It is the new ipsec session with same/reused ip + same user token that is a bit more tricky for the application to detect as "new", since it has the same token+ip as the previous ipsec session.
> 
> Why does it have to be detected as new? I mean it's the same token and IP.

In short: If the application is unaware of that a client has dropped out it will keep a dedicated ip socket and session open. If the (same) client reconnects to that open ip socket there is a risk of confusion between the server application and the client about old/new validity of their session state/status of the application SA, causing unnecessary reconnects/retries/timeouts. This is completely avoided if a client always comes in with a new ip. (The application protocol definitely has its flaws, but it is set in stone and out of my control.)

**Reply by tobiasbrunner** [reply]
> I suppose that the dhcp plugin will only ask the dhcpd server for an initial ip address (with DHCPDISCOVER) when the client asks for an ip address, and that won't happen upon rekeying only. Or will it?

Yeah, leases are currently not renewed automatically (the plugin doesn't even parse the lease time option). But the lease is released when an IKE_SA goes down.
